### ISSUE TITLE
Run workflow steps do not inherit environment configuration from parents

### ISSUE THREAD

**Issue by cirego** [MaterializeInc/materialize#5588]
I have an example workflow that sets some environment variables:

```yaml
  ingest-bench-ci:
    env:
      MZ_WORKERS: 4
      CHBENCH_RUN_SECONDS: 60
      NUM_WAREHOUSES: 1
      OLTP_THREADS: 4
    steps:
      - step: workflow
        workflow: generate-snapshot
```

Via some intermediate steps, this calls `heavy-load-mysql`. However, it appears that variables are instead being populated from the environment:

```yaml
  heavy-load-mysql:
    steps:
    - step: run
      service: chbench
      daemon: true
      command: >-
        run
        --dsn=mysql --gen-dir=/var/lib/mysql-files
        --analytic-threads=0
        --transactional-threads=${OLTP_THREADS:-1}
        --run-seconds=${CHBENCH_RUN_SECONDS:-864000}
        -l /var/log/chbench.log
        --config-file-path=/etc/chbenchmark/mz-default-mysql.cfg
        --mz-url=postgresql://${MZ_HOST:-materialized}:6875/materialize?sslmode=disable
```

When I run `ingest-bench-ci`, `docker inspect` shows that `chbench` is running with
`--run-seconds=864000`:

```
            "Cmd": [
                "run",
                "--dsn=mysql",
                "--gen-dir=/var/lib/mysql-files",
                "--analytic-threads=0",
                "--transactional-threads=1",
                "--run-seconds=864000",
                "-l",
                "/var/log/chbench.log",
                "--config-file-path=/etc/chbenchmark/mz-default-mysql.cfg",
                "--mz-url=postgresql://materialized:6875/materialize?sslmode=disable"
            ],
```

Ideally, when running a workflow, we would propogate any environment variables downwards to
sub-workflows.

benesch or @quodlibetor, would love to hear you thoughts on this.

**Comment by cirego** [comment]
@benesch did you have thoughts on this?

I thought I had tagged you, but it appears I missed the `@`.

**Comment by cirego** [comment]
I have currently have this as a blocker for writing a CI job to verify the chbench ingest benchmark.

**Comment by benesch** [comment]
I'm afraid I don't have a good handle on how the environment variables in the `mzworkflows` block interacts with the environment variables in the rest of the Docker Compose file. Personally I find environment variables really hard to reason about: it's hard to find out where they've been set; it's easy to accidentally clear out the environment when forking child processes; they're not resilient to typos; etc.

In my ideal world we'd not use environment variables in mzcompose at all, except as hardcoded configuration parameters for containers where required. 

Roughly, I was thinking it'd be sweet if we allowed mzcompose workflows to define flags that could be plumbed around explicitly, like so:

```yml
  ingest-bench-ci:
    flags:
      - mz-workers: 4
      - chbench-run-seconds: 60
      - num-warehouses: 1
      - oltp-threads: 4
    steps:
      - step: workflow
        workflow: heavy-load-mysql
        flags:
          mz-workers: %{mz-workers}
          chbench-run-seconds: %{chbench-run-seconds}
          num-warehouses: %{num-warehouses}
          oltp-threads: %{oltp-threads}
  heavy-load-mysql:
    flags:
      - mz-workers
      - chbench-run-seconds
      - num-warehouses
      - oltp-threads
    steps:
    - step: run
      service: chbench
      daemon: true
      command: >-
        run
        --dsn=mysql --gen-dir=/var/lib/mysql-files
        --analytic-threads=0
        --transactional-threads=%{oltp-threads}
        --run-seconds=%{chbench-run-seconds}
        -l /var/log/chbench.log
        --config-file-path=/etc/chbenchmark/mz-default-mysql.cfg
        --mz-url=postgresql://%{mz-host}:6875/materialize?sslmode=disable
```

**Comment by benesch** [comment]
Then you'd specify those flags as e.g.:

```
./mzcompose run ingest-bench-ci -- --chbench-run-seconds=80
```

and if a flag didn't have a default it'd be required.

That's a pretty heavy lift though. Maybe there's something simpler we could do in the meantime, but like I said I find the environment variable semantics confusing so I'm probably not the right guy to figure out what that simpler thing is.

I think I've mentioned this elsewhere, and it's probably a bit separate, but I'd be delighted if these workflows supported a Python script escape hatch. A lot of this stuff gets awkward in YAML fast. If we instead allowed `./mzcompose run SOME-FOO` to call some Python script called `SOME-FOO.py`, that thing could do whatever argument parsing etc. it wanted without having to make architectural changes to mzcompose. Roughly it'd be nice if the workflow steps were also exposed as a Python library, so you could write stuff like:

```py
# SOME-FOO.py

def main(composition: mzcompose.Composition):
    parser = argparse.ArgumentParser(description='Blah blah blah.')
    # ...
    args = parser.parse_args()

    composition.start_services("blah")
    if args.blah:
        do_complicated_blah_thing()
    composition.stop_services("blah")
```

So this is a bit of a stream of consciousness. Let me know if anything doesn't make sense. A simpler solution that just unblocks you seems like a fine place for now.

**Comment by cirego** [comment]
The stream of consciousness notes are great! I've been wanting mzcompose to be a library for a while now -- and while this might be a big lift, I think it's worth doing.

I have another PR that wraps `mzcompose` to run a matrix benchmark and it's passing arguments into `mzcompose` using environment variables. It seems kind of silly to have these layers of scripts that are taking arguments, parsing them, passing them down the chain... repeat.

Teaching `mzcompose` to take a flags block, and parse flags from the CLI, would certainly help, but I feel like this is all working around the fact that we really want to be writing programs for workflows. I hadn't thought about the possibility of having `mzcompose` take a script as a workflow, either explicitly or implicitly, as an argument.

**Comment by benesch** [comment]
> we really want to be writing programs for workflows.

100% agree. Certainly for the workloads of any notable complexity.

>  I hadn't thought about the possibility of having mzcompose take a script as a workflow, either explicitly or implicitly, as an argument.

Yeah, I was thinking that'd be the easiest way to mash up the YAML and Python worlds. I actually remember now that I started mocking up this idea like so:

```py
# mzworkflows.py

def workflow_one(composition: mzcompose.Composition, args: Iterable[str]):
    pass

def workflow_two(composition: mzcompose.Composition, args: Iterable[str]):
    pass
```

The idea being you could drop an `mzworkflows.py` file into any directory with an mzcompose.yml file. Then mzcompose would use `importlib.import_module` to attempt to load that file. Any functions in that file whose name began with `workflow_` and took the right kind of parameters would be considered to be workflows, just like workflows defined in YAML. You'd be able to call them directly with `./mzcompose run`, or use them from a YAML WorkflowWorkflowStep, etc.

**Comment by cirego** [comment]
This issue was fixed by #5925. Going to file a new issue to track the `mzworkflows.py` idea!

**Comment by benesch** [comment]
👍🏽 assuming that you meant to close this!

**Comment by cirego** [comment]
> 👍🏽 assuming that you meant to close this!

Yes, thank you!
